### ISSUE TITLE
`GitHubRepo._clone_repo()` should fetch updates to cached local repos if local repo already exists

### ISSUE THREAD

**Issue by andyk** [agentos-project/agentos#310]
I noticed today after I pushed changes to the `test_staging` branch in the main agentos-project/agentos repo that those changes weren't getting pulled down by AgentOS automatically.

I think I found the problem in `GitHubRepo._clone_repo()`:
https://github.com/agentos-project/agentos/blob/54928a779e673c52e816c2bf84b675a597e45640/agentos/repo.py#L319-L330

If a local repo exists already, this function should call `porcelain.fetch()` on so that it gets updates form the remote.

**Comment by andyk** [comment]
I thought I had fixed this by adding a `porcelain.fetch()`  as in:

https://github.com/agentos-project/agentos/blob/9316319a6e1ccb03b3c349f20459b8fd39dc88a1/agentos/repo.py#L339-L347

But then when I ran that code in a debugger, I saw that command did run as:
```
porcelain.fetch(
    repo='/Users/andyk/.agentos/cache/repos_cache/agentos-project/agentos.git/test_staging',
    remote_location='https://github.com/agentos-project/agentos.git')
)
```
...and it didn't throw an exception but also didn't actually perform a `git fetch` in the repo at `/Users/andyk/.agentos/cache/repos_cache/agentos-project/agentos.git/test_staging`.

I determined this by running:

```
$ cd /Users/andyk/.agentos/cache/repos_cache/agentos-project/agentos.git/test_staging
$ git fetch 
From https://github.com/agentos-project/agentos
   430344b..54928a7  master       -> origin/master
 + 2f830d5...9316319 test_staging -> origin/test_staging  (forced update)
   f191051..1a8803f  website      -> origin/website
 * [new tag]         v0.1.0       -> v0.1.0
```

You can see from the output that the remote branches were not actually fetched when my `porcelain.fetch()` command ran.

So I guess I'm probably using dulwich/porcelain fetch incorrectly.

@nickjalbert any insights?

**Comment by nickjalbert** [comment]
This was closed automatically when I merged your PR, but I'm not sure it should have been (or, at least, we should confirm the bug is fixed).  

I also ran into problems similar to this when swapping git for dulwich.  I'm still not sure if I misunderstood the porcelain API or the git internals, but I did spend a chunk of time looking at the dulwich source to trying to figure it out.  Finally, I timeboxed it and decided to return to it in the future.

Long story short, I found dulwich a bit hard to wrangle and I don't feel I have a great grasp on it.

About your specific question, I would have expected the fetch to work (note: I think *then* you'd need to update HEAD to the latest fetched version of your branch to get the behavior you wanted) so, alas, not much help there. 😭

**Comment by andyk** [comment]
Yeah the code also does the reset --hard after the fetch.

I was in a very similar situation yesterday to what you described where I ended up time boxing my digging into dulwich code. Worst case, for now we can delete local repos when we suspect the issue is with caches.

We should just keep in mind that other people might run into bugs like this.

**Comment by andyk** [comment]
This is definitely still a bug. I just run into this bug again when I tried running tests after merging your PR which switched the testing branch back to `test_prod`. I got an error as though my tests were running against an old version of that branch, so I deleted my cached agentos-project repo and re-ran my tests and they passed.

**Comment by nickjalbert** [comment]
It's probably not a bad idea for the tests to clear all caches (or to not use any caches).  Potentially a separate issue from this...

**Comment by andyk** [comment]
Good suggestion @nickjalbert, just created #329 to track it

**Comment by andyk** [comment]
Clearing these caches can be tricky if modules that live in them have already been loaded into a python process. Just ran into this when trying to update the tests to clear the repo cache, but then Windows (though not Linux or MacOs) threw an exception during `test_component_from_github_no_venv` when trying to clear the cache because `test_component_from_github_with_venv` had already run and loaded some modules from that same local github repo.

I now think perhaps we should only cache repos locally into directories that have the name of the github hash, and then we can also separately cache the mapping between branch names and repos.

**Comment by nickjalbert** [comment]
To record our discussion from yesterday:

> cache repos locally into directories that have the name of the github hash

I dig this!

Another thing we could do is to set the cache directory to a `tmpdir` in the tests so there will be no reuse between tests.

**Comment by nickjalbert** [comment]
https://github.com/agentos-project/agentos/pull/379 should fix this.

**Comment by nickjalbert** [comment]
Fixed in #379
